File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/domene/bransjeprogram/Bransje.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.bransjeprogram;

import java.util.Arrays;
import java.util.List;
import lombok.Data;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.Næringskode5Siffer;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domene.UnderenhetLegacy;

@Data
public class Bransje {

  public final ArbeidsmiljøportalenBransje type;
  public final String navn;
  public final List<String> koderSomSpesifisererNæringer;

  public Bransje(
      ArbeidsmiljøportalenBransje type, String navn, String... koderSomSpesifisererNæringer) {
    this.type = type;
    this.navn = navn;
    this.koderSomSpesifisererNæringer = Arrays.asList(koderSomSpesifisererNæringer);
    validerKoder();
  }

  private void validerKoder() {
    if (inneholderKunKoderMedGittAntallSifre(2) || inneholderKunKoderMedGittAntallSifre(5)) {
      return;
    }
    throw new IllegalArgumentException(
        "Støtter kun bransjer som er spesifisert av enten 2 eller 5 sifre");
  }

  private boolean inneholderKunKoderMedGittAntallSifre(int antallSifre) {
    return koderSomSpesifisererNæringer.stream().allMatch(kode -> kode.length() == antallSifre);
  }

  public boolean erDefinertPåTosiffernivå() {
    return inneholderKunKoderMedGittAntallSifre(2);
  }

  public boolean erDefinertPåFemsiffernivå() {
    return inneholderKunKoderMedGittAntallSifre(5);
  }

  public boolean inkludererVirksomhet(UnderenhetLegacy underenhet) {
    return inkludererNæringskode(underenhet.getNæringskode());
  }

  public boolean inkludererNæringskode(Næringskode5Siffer næringskode5Siffer) {
    String næringskode = næringskode5Siffer.getKode();
    return koderSomSpesifisererNæringer.stream().anyMatch(næringskode::startsWith);
  }

  public boolean inkludererNæringskode(String næringskode5Siffer) {
    if (næringskode5Siffer == null) {
      return false;
    }
    return koderSomSpesifisererNæringer.stream().anyMatch(næringskode5Siffer::startsWith);
  }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/domenemodeller/bransjeprogram/Bransje.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.bransjeprogram

import ia.felles.definisjoner.bransjer.Bransjer
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.Næringskode
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.Virksomhet

data class Bransje(
    val type: Bransjer,
) {
    val navn = type.navn
    val identifikatorer = type.næringskoder

    init {
        require(erDefinertPåTosiffernivå() || erDefinertPåFemsiffernivå())
        { "Støtter kun bransjer som er spesifisert av enten 2 eller 5 sifre" }
    }

    fun erDefinertPåTosiffernivå() = identifikatorer.all { it.length == 2 }

    fun erDefinertPåFemsiffernivå() = identifikatorer.all { it.length == 5 }

    fun inkludererVirksomhet(underenhet: Virksomhet): Boolean {
        return inkludererNæringskode(underenhet.næringskode)
    }

    fun inkludererNæringskode(næringskode: Næringskode?): Boolean {
        if (næringskode == null) {
            return false
        }
        return identifikatorer.any {
            næringskode.femsifferIdentifikator.startsWith(it)
        }
    }
}
